package/MDAnalysis/analysis/base.py

```diff
@@ -125,14 +125,105 @@
+from functools import partial
 
 logger = logging.getLogger(__name__)
 
+class ParallelExecutor(object):
+    """Class that executes parallel computations for AnalysisBase (but potentially for other purposes)
+    using one of the available backends"""
+
+    @classmethod
+    @property
+    def possible_backends(cls):
+        return ("local", "dask", "multiprocessing", "dask.distributed")
+
+    def __init__(self, backend: str, n_workers: int, client: object = None):
+        # validate 'client' argument
+        if client is not None and backend is not None:
+            raise ValueError(
+                (f"If 'client' is not None, backend is deduced and must be None, but here "
+                f'{backend=}, {client=}')
+            )
+        if client is not None:
+            # assuming it's a dask distributed client
+            from dask.distributed import Client as DistributedClient
+            if not isinstance(client, DistributedClient):
+                raise ValueError(f"client should be dask.distributed.Client instance, got {type(client)}")
+        self.client = client
+
+        # validate 'backend' argument
+        if backend is not None:
+            if backend not in ParallelExecutor.possible_backends:
+                raise ValueError(
+                    f"Invalid {backend=}: should be one of {ParallelExecutor.possible_backends}"
+                )
+            if backend == "local":
+                if n_workers > 1:
+                    raise ValueError("Can be only 1 worker with backend='local', got {n_workers=")
+            else:
+                try:
+                    import_module(backend)
+                except ImportError:
+                    raise ValueError(
+                        "Please install {backend} module "
+                        "with 'python3 -m pip install {backend} "
+                        "to run this")
+        self.backend = backend
+
+        # validate 'n_workers' argument
+        if n_workers and n_workers < 0:
+            raise ValueError(f"n_workers should be non-negative, got {n_workers=}")
+        self.n_workers = n_workers
+
+    def apply(self, func: Callable, computations: list) -> list:
+        if self.backend == "local":
+            return self._compute_with_local(func, computations)
+        elif self.backend == "multiprocessing":
+            return self._compute_with_multiprocessing(func, computations)
+        elif self.backend == "dask":
+            return self._compute_with_dask(func, computations)
+        elif self.client:
+            return self._compute_with_client(func, computations)
+        elif self.backend == 'dask.distributed':
+            raise ValueError(f"To use this backend, pre-configure a Client object and pass client=... instead")
+        else:
+            raise ValueError(f"Backend or client is not set properly: {self.backend=}, {self.client=}")
+
+    def _compute_with_local(self, func: Callable, computations: list) -> list:
```

Comment:
docstrings on all these please, otherwise it's impossible tto know what they are intended to do

Reply:
done